plugins/modules/mssql_script.py

```diff
@@ -238,14 +288,106 @@ def clean_output(o):
     return str(o)
 
 
+def detect_and_remove_bom(content):
+    """
+    Detect and remove Byte Order Mark (BOM) from content.
+    Returns tuple (cleaned_content, bom_removed, bom_type)
+    """
+    bom_removed = False
+    bom_type = None
+
+    # Common BOM patterns
+    boms = [
+        (codecs.BOM_UTF8, 'UTF-8'),
+        (codecs.BOM_UTF16_LE, 'UTF-16 LE'),
+        (codecs.BOM_UTF16_BE, 'UTF-16 BE'),
+        (codecs.BOM_UTF32_LE, 'UTF-32 LE'),
+        (codecs.BOM_UTF32_BE, 'UTF-32 BE'),
+    ]
+
+    # Check if content is bytes or string
+    if isinstance(content, str):
+        content_bytes = content.encode('utf-8')
+    else:
+        content_bytes = content
+
+    # Check for BOM
+    for bom, bom_name in boms:
+        if content_bytes.startswith(bom):
+            content_bytes = content_bytes[len(bom):]
+            bom_removed = True
+            bom_type = bom_name
+            break
+
+    # Convert back to string if original was string
+    if isinstance(content, str):
+        try:
+            cleaned_content = content_bytes.decode('utf-8')
+        except UnicodeDecodeError:
+            # Fallback to original encoding detection
+            cleaned_content = content_bytes.decode('utf-8', errors='replace')
+    else:
+        cleaned_content = content_bytes
+
+    return cleaned_content, bom_removed, bom_type
+
+
+def read_script_file(script_path):
+    """
+    Read SQL script file and remove BOM if present.
+    Returns tuple (script_content, bom_info)
+    """
+    if not os.path.exists(script_path):
+        raise IOError(f"Script file not found: {script_path}")
+
+    if not os.path.isfile(script_path):
+        raise IOError(f"Path is not a file: {script_path}")
+
+    try:
+        with open(script_path, 'rb') as f:
+            raw_content = f.read()
+
+        # Convert to string for processing
+        try:
+            script_content = raw_content.decode('utf-8')
+        except UnicodeDecodeError:
+            # Try other common encodings
+            for encoding in ['utf-16', 'latin1', 'cp1252']:
+                try:
+                    script_content = raw_content.decode(encoding)
+                    break
+                except UnicodeDecodeError:
+                    continue
+            else:
+                # Last resort - decode with error replacement
+                script_content = raw_content.decode('utf-8', errors='replace')
+
+        # Detect and remove BOM
+        cleaned_content, bom_removed, bom_type = detect_and_remove_bom(
```

Comment:
Shouldn't the BOM detection/removal happen **before** any decoding is done? The BOM basically tells you the encoding you have to use to decode the script.